lib/HLSL/DxilCondenseResources.cpp

```diff
@@ -2258,17 +2283,61 @@ static void MarkCBUse(unsigned offset, FieldAnnotationByOffsetMap &fieldMap) {
     it->second->SetCBVarUsed(true);
 }
 
-static unsigned GetOffsetForCBExtractValue(ExtractValueInst *EV, bool bMinPrecision) {
+// Detect patterns of lshr v,16 or trunc to 16-bits and return low and high
+// word usage.
+static const unsigned kLowWordUsed = 1;
+static const unsigned kHighWordUsed = 2;
+static const unsigned kLowHighWordMask = kLowWordUsed | kHighWordUsed;
+static unsigned DetectLowAndHighWordUsage(ExtractValueInst *EV) {
+  unsigned result = 0;
+  if (EV->getType()->getScalarSizeInBits() == 32) {
+    for (auto U : EV->users()) {
+      Instruction *I = cast<Instruction>(U);
+      if (I->getOpcode() == LLVMLShr) {
+        ConstantInt *CShift = dyn_cast<ConstantInt>(I->getOperand(1));
+        if (CShift && CShift->getLimitedValue() == 16)
+          result |= kHighWordUsed;
+      } else if (I->getOpcode() == LLVMTrunc &&
+               I->getType()->getPrimitiveSizeInBits() == 16) {
+        result |= kLowWordUsed;
+      } else {
+        // Assume whole dword is used, return 0
+        return 0;
```

Comment:
What is the difference between return 0 and return kLowHighWordMask?

Reply:
If 0 is returned, it will mark only one value as used at beginning of offset because the it doesn't appear to be using high and low as separate 16-bit values.  0 is also returned for the case where the pattern doesn't apply at all because the value size wasn't 32-bits, so you only want to go down the same path as before and mark only one value.  If kLowHighWordMask (both bits set) is returned, then two separate 16-bit values are being used, so it will mark two values, one at the beginning, and one offset by 2 bytes.  So returning 0 results in the same behavior as returning only kLowWordUsed for now, but the meaning is different, so mixing them up could be confusing.